File: app/src/main/java/com/example/studybuddy/AddTaskActivity.java
package com.example.studybuddy; // Asegúrate de que este sea tu package name

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import com.google.android.material.textfield.TextInputEditText;

import com.example.studybuddy.notification.AlarmManagerHelper; // *** Importación para la alarma ***

import adapter.TaskAdapter;
import model.Task;
import viewmodel.TaskViewModel;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

public class AddTaskActivity extends AppCompatActivity {

    public static final String EXTRA_TASK_ID = "extra_task_id";
    public static final String EXTRA_TASK_TITLE = "extra_task_title";
    public static final String EXTRA_TASK_DESCRIPTION = "extra_task_description";
    public static final String EXTRA_TASK_DATE = "extra_task_date";
    public static final String EXTRA_TASK_TIME = "extra_task_time";
    public static final String EXTRA_TASK_COMPLETED = "extra_task_complete";
    private TextInputEditText editTextTitle, editTextDescription;
    private Button buttonSelectDate, buttonSelectTime, buttonSaveTask;
    private TextView textViewSelectedDate, textViewSelectedTime, textViewAddTitle;


    private TaskViewModel taskViewModel;

    private Calendar selectedDateCalendar;
    private SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
    private SimpleDateFormat timeFormat = new SimpleDateFormat("HH:mm", Locale.getDefault());

    private int taskId = -1;
    private boolean isEditing = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_task);

        initViews();
        setupViewModel();
        setupClickListeners();
        checkForEditIntent(); // Verificar si es edición o nueva tarea
    }

    private void initViews() {
        editTextTitle = findViewById(R.id.edit_text_title);
        editTextDescription = findViewById(R.id.edit_text_description);
        textViewSelectedDate = findViewById(R.id.text_view_selected_date);
        textViewSelectedTime = findViewById(R.id.text_view_selected_time);
        buttonSelectDate = findViewById(R.id.button_select_date);
        buttonSelectTime = findViewById(R.id.button_select_time);
        buttonSaveTask = findViewById(R.id.button_save_task);
        textViewAddTitle = findViewById(R.id.text_view_add_title);

    }

    private void setupViewModel() {
        // Asegúrate de que TaskViewModel esté correctamente definido y en el paquete correcto
        taskViewModel = new ViewModelProvider(this).get(TaskViewModel.class);
    }

    private void setupClickListeners() {
        buttonSelectDate.setOnClickListener(v -> showDatePickerDialog());
        buttonSelectTime.setOnClickListener(v -> showTimePickerDialog());
        buttonSaveTask.setOnClickListener(v -> saveTask());
    }

    private void checkForEditIntent() {
        Intent intent = getIntent();
        if (intent.hasExtra(EXTRA_TASK_ID)) {
            // Es una edición
            isEditing = true;
            taskId = intent.getIntExtra(EXTRA_TASK_ID, -1);
            String title = intent.getStringExtra(EXTRA_TASK_TITLE);
            String description = intent.getStringExtra(EXTRA_TASK_DESCRIPTION);
            String date = intent.getStringExtra(EXTRA_TASK_DATE);
            String time = intent.getStringExtra(EXTRA_TASK_TIME);

            editTextTitle.setText(title != null ? title : "");
            editTextDescription.setText(description != null ? description : "");

            // Parsear fecha y hora para el Calendar
            try {
                String dateTimeString = date + " " + time;
                SimpleDateFormat fullFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault());
                Date parsedDate = fullFormat.parse(dateTimeString);
                if (parsedDate != null) {
                    selectedDateCalendar = Calendar.getInstance();
                    selectedDateCalendar.setTime(parsedDate);
                } else {
                    selectedDateCalendar = Calendar.getInstance();
                }
            } catch (ParseException e) {
                e.printStackTrace();
                selectedDateCalendar = Calendar.getInstance();
            }
            updateDateAndTimeViews();
            buttonSaveTask.setText("Actualizar Tarea");
            TextView textViewAddTitle = findViewById(R.id.text_view_add_title);
            textViewAddTitle.setText("Editar Tarea");
        } else {
            // Es una nueva tarea
            isEditing = false;
            taskId = -1;
            selectedDateCalendar = Calendar.getInstance(); // *** Corregido ***
            updateDateAndTimeViews();
            buttonSaveTask.setText("Guardar Tarea");
            TextView textViewAddTitle = findViewById(R.id.text_view_add_title);
            textViewAddTitle.setText("Agregar Tarea");
        }
    }

    private void showDatePickerDialog() {
        if (selectedDateCalendar == null) selectedDateCalendar = Calendar.getInstance();
        DatePickerDialog datePickerDialog = new DatePickerDialog(
                this,
                (view, year, month, dayOfMonth) -> {
                    selectedDateCalendar.set(Calendar.YEAR, year);
                    selectedDateCalendar.set(Calendar.MONTH, month);
                    selectedDateCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                    updateDateAndTimeViews();
                },
                selectedDateCalendar.get(Calendar.YEAR),
                selectedDateCalendar.get(Calendar.MONTH),
                selectedDateCalendar.get(Calendar.DAY_OF_MONTH)
        );
        datePickerDialog.show();
    }

    private void showTimePickerDialog() {
        if (selectedDateCalendar == null) selectedDateCalendar = Calendar.getInstance();
        TimePickerDialog timePickerDialog = new TimePickerDialog(
                this,
                (view, hourOfDay, minute) -> {
                    selectedDateCalendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
                    selectedDateCalendar.set(Calendar.MINUTE, minute);
                    updateDateAndTimeViews();
                },
                selectedDateCalendar.get(Calendar.HOUR_OF_DAY),
                selectedDateCalendar.get(Calendar.MINUTE),
                true // Formato 24 horas
        );
        timePickerDialog.show();
    }

    private void updateDateAndTimeViews() {
        if (selectedDateCalendar != null) {
            textViewSelectedDate.setText(dateFormat.format(selectedDateCalendar.getTime()));
            textViewSelectedTime.setText(timeFormat.format(selectedDateCalendar.getTime()));
        } else {
            textViewSelectedDate.setText("Seleccionar Fecha");
            textViewSelectedTime.setText("Seleccionar Hora");
        }
    }

    private void saveTask() {
        String title = editTextTitle.getText().toString().trim();
        String description = editTextDescription.getText().toString().trim();
        String date = textViewSelectedDate.getText().toString();
        String time = textViewSelectedTime.getText().toString();

        if (title.isEmpty()) {
            editTextTitle.setError("El título es obligatorio");
            editTextTitle.requestFocus();
            return;
        }

        // *** VERIFICAR SI LA FECHA Y HORA ESTÁN VACÍAS ***
        if (date.isEmpty() || time.isEmpty()) {
            Toast.makeText(this, "Por favor, seleccione una fecha y hora válidas", Toast.LENGTH_SHORT).show();
            return; // No continuar si falta fecha/hora
        }

        // *** INICIO: LÓGICA PARA DETERMINAR SI ES EDICIÓN O NUEVA TAREA ***
        if (isEditing) {
            // *** ESCENARIO: Editar tarea existente ***
            Task taskToUpdate = new Task(title, description, date, time, false);
            taskToUpdate.setId(taskId);
            taskViewModel.update(taskToUpdate);

            // *** ALARMA: Reprogramar la alarma para la tarea actualizada ***
            // Llamamos al método auxiliar
            scheduleAlarmForTask(taskToUpdate, date, time);

            Toast.makeText(this, "Tarea actualizada", Toast.LENGTH_SHORT).show();
        } else {
            // *** ESCENARIO: Crear nueva tarea ***
            Task newTask = new Task(title, description, date, time, false);
            taskViewModel.insert(newTask);

            // *** ALARMA: Programar la alarma para la nueva tarea ***
            // Llamamos al método auxiliar
            scheduleAlarmForTask(newTask, date, time);

            Toast.makeText(this, "Tarea guardada", Toast.LENGTH_SHORT).show();
        }
        // *** FIN: LÓGICA DE GUARDADO Y ALARMA ***

        finish();
    }

    // *** MÉTODO AUXILIAR PARA PROGRAMAR LA ALARMA ***
    /**
     * Programa una alarma para una tarea específica usando AlarmManagerHelper.
     * @param task El objeto Task (debe tener un ID asignado si es nueva, o el ID existente si es editada).
     * @param dateString La fecha en formato String (esperado "yyyy-MM-dd").
     * @param timeString La hora en formato String (esperado "HH:mm").
     */
    private void scheduleAlarmForTask(Task task, String dateString, String timeString) {
        try {
            // *** PARSEAR FECHA Y HORA ***
            // Asumiendo formato "yyyy-MM-dd" para date y "HH:mm" para time
            String[] dateParts = dateString.split("-");
            int year = Integer.parseInt(dateParts[0]);
            // ¡OJO! Los meses en Calendar van de 0 (Enero) a 11 (Diciembre)
            int month = Integer.parseInt(dateParts[1]) - 1;
            int day = Integer.parseInt(dateParts[2]);

            String[] timeParts = timeString.split(":");
            int hour = Integer.parseInt(timeParts[0]);
            int minute = Integer.parseInt(timeParts[1]);

            // Validar rangos razonables
            if (year < 2000 || year > 2100 || month < 0 || month > 11 || day < 1 || day > 31 ||
                    hour < 0 || hour > 23 || minute < 0 || minute > 59) {
                throw new IllegalArgumentException("Fecha/Hora fuera de rango válido");
            }

            // *** USO DE ALARM MANAGER HELPER PARA PROGRAMAR LA ALARMA ***
            AlarmManagerHelper alarmManagerHelper = new AlarmManagerHelper((TaskAdapter.OnItemClickListener) AddTaskActivity.this);
            alarmManagerHelper.setAlarm(
                    task.getId(),
                    task.getTitle(),
                    task.getDescription(),
                    year,
                    month,
                    day,
                    hour,
                    minute
            );

        } catch (NumberFormatException e) {
            // Manejar errores de parsing (formato incorrecto, valores no numéricos)
            Log.e("AddTaskActivity", "Error al parsear fecha/hora: " + e.getMessage());
            Toast.makeText(this, "Error al programar recordatorio: Fecha/Hora inválida", Toast.LENGTH_SHORT).show();
        } catch (IllegalArgumentException e) {
            // Manejar errores de rango
            Log.e("AddTaskActivity", "Error al validar fecha/hora: " + e.getMessage());
            Toast.makeText(this, "Error al programar recordatorio: Fecha/Hora inválida", Toast.LENGTH_SHORT).show();
        } catch (Exception e) {
            // Capturar otros errores inesperados
            Log.e("AddTaskActivity", "Error inesperado al programar alarma: " + e.getMessage());
            Toast.makeText(this, "Error al programar recordatorio: Fecha/Hora inválida", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/studybuddy/AddTaskActivity.kt
package com.example.studybuddy

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.res.ColorStateList
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import com.example.studybuddy.databinding.ActivityAddTaskBinding
import com.example.studybuddy.model.CategoryEntity
import com.example.studybuddy.model.Priority
import com.example.studybuddy.model.RepeatInterval
import com.example.studybuddy.model.Task
import com.example.studybuddy.model.categoryIconRes
import com.example.studybuddy.model.displayName
import com.example.studybuddy.notification.AlarmManagerHelper
import com.example.studybuddy.ui.TaskViewModel
import com.example.studybuddy.util.DateUtils
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.util.Calendar

class AddTaskActivity : AppCompatActivity() {

    companion object {
        const val EXTRA_TASK_ID = "extra_task_id"
    }

    private lateinit var binding: ActivityAddTaskBinding
    private lateinit var taskViewModel: TaskViewModel
    private lateinit var alarmHelper: AlarmManagerHelper

    private var taskId = -1
    private var isEditing = false
    private var selectedCategory = CategoryEntity.ID_GENERAL
    private var selectedPriority = Priority.MEDIUM
    private var selectedRepeat = RepeatInterval.NONE
    private var selectedDateCalendar = Calendar.getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddTaskBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        taskViewModel = ViewModelProvider(this)[TaskViewModel::class.java]
        alarmHelper = AlarmManagerHelper(this)

        setupCategoryChips()
        setupPriorityChips()
        setupRepeatChips()
        setupClickListeners()
        checkForEditIntent()
        updateDateAndTimeViews()
    }

    private fun setupCategoryChips() {
        binding.chipCategoryGeneral.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedCategory = CategoryEntity.ID_GENERAL
        }
        binding.chipCategoryStudy.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedCategory = CategoryEntity.ID_STUDY
        }
        binding.chipCategoryWork.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedCategory = CategoryEntity.ID_WORK
        }
        binding.chipCategoryPersonal.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedCategory = CategoryEntity.ID_PERSONAL
        }
        binding.chipCategoryNew.setOnClickListener { showCreateCategoryDialog() }
        taskViewModel.categories.observe(this) { categories ->
            renderCategoryChips(categories)
        }
    }

    private fun renderCategoryChips(categories: List<CategoryEntity>) {
        val existing = mutableMapOf<Int, Chip>()
        for (i in 0 until binding.chipGroupCategory.childCount) {
            val chip = binding.chipGroupCategory.getChildAt(i) as? Chip ?: continue
            val id = chip.tag as? Int ?: continue
            existing[id] = chip
        }
        val seen = mutableSetOf<Int>()
        categories.filter { it.id !in CategoryEntity.ID_GENERAL..CategoryEntity.ID_PERSONAL }
            .forEach { category ->
                seen.add(category.id)
                if (existing[category.id] != null) return@forEach
                val iconRes = category.icon.categoryIconRes() ?: R.drawable.ic_cat_star
                val chip = Chip(this).apply {
                    tag = category.id
                    text = category.displayName(this@AddTaskActivity)
                    chipIcon = ContextCompat.getDrawable(this@AddTaskActivity, iconRes)
                    chipIconTint = ColorStateList.valueOf(category.color)
                    isCheckable = true
                    setOnCheckedChangeListener { _, isChecked ->
                        if (isChecked) selectedCategory = category.id
                    }
                    setOnLongClickListener {
                        confirmDeleteCategory(category)
                        true
                    }
                }
                binding.chipGroupCategory.addView(chip)
            }
        existing.forEach { (id, chip) ->
            if (id !in seen) {
                binding.chipGroupCategory.removeView(chip)
            }
        }
    }

    private fun showCreateCategoryDialog() {
        val inflater = LayoutInflater.from(this)
        val dialogView = inflater.inflate(
            R.layout.dialog_new_category, null
        ) as View
        val nameInput = dialogView.findViewById<EditText>(R.id.edit_text_category_name)
        val iconGroup = dialogView.findViewById<ChipGroup>(R.id.chip_group_category_icon)

        CategoryEntity.ICON_KEYS.forEachIndexed { index, iconKey ->
            val iconRes = iconKey.categoryIconRes() ?: return@forEachIndexed
            val chip = Chip(this).apply {
                text = " "
                chipIcon = ContextCompat.getDrawable(this@AddTaskActivity, iconRes)
                chipIconTint = ColorStateList.valueOf(
                    android.graphics.Color.parseColor("#616161")
                )
                isCheckable = true
                isChecked = index == 0
            }
            chip.tag = iconKey
            iconGroup.addView(chip)
        }

        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.new_category)
            .setView(dialogView)
            .setPositiveButton(R.string.save) { dialog, _ ->
                val name = nameInput.text.toString().trim()
                if (name.isEmpty()) {
                    nameInput.error = getString(R.string.category_name_required)
                    return@setPositiveButton
                }
                val icon = (iconGroup.checkedChipId.let {
                    iconGroup.findViewById<Chip>(it)?.tag as? String
                }) ?: CategoryEntity.DEFAULT_ICON
                val customCount = (taskViewModel.categories.value ?: emptyList())
                    .count { it.id !in CategoryEntity.ID_GENERAL..CategoryEntity.ID_PERSONAL }
                val color = CategoryEntity.CUSTOM_COLORS[
                    customCount % CategoryEntity.CUSTOM_COLORS.size
                ]
                taskViewModel.upsertCategory(
                    CategoryEntity(name = name, icon = icon, color = color)
                ) { id ->
                    selectCategoryChip(id)
                }
                dialog.dismiss()
            }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    private fun confirmDeleteCategory(category: CategoryEntity) {
        MaterialAlertDialogBuilder(this)
            .setTitle(R.string.delete_category_title)
            .setMessage(getString(R.string.delete_category_message, category.displayName(this)))
            .setPositiveButton(R.string.delete) { _, _ ->
                taskViewModel.deleteCategory(category)
                selectedCategory = CategoryEntity.ID_GENERAL
                binding.chipCategoryGeneral.isChecked = true
            }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    private fun selectCategoryChip(categoryId: Int) {
        for (i in 0 until binding.chipGroupCategory.childCount) {
            val chip = binding.chipGroupCategory.getChildAt(i) as? Chip ?: continue
            if (chip.tag == categoryId) {
                chip.isChecked = true
                selectedCategory = categoryId
                return
            }
        }
    }

    private fun setupPriorityChips() {
        binding.chipPriorityLow.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedPriority = Priority.LOW
        }
        binding.chipPriorityMedium.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedPriority = Priority.MEDIUM
        }
        binding.chipPriorityHigh.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedPriority = Priority.HIGH
        }
    }

    private fun setupRepeatChips() {
        binding.chipRepeatNone.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedRepeat = RepeatInterval.NONE
        }
        binding.chipRepeatDaily.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedRepeat = RepeatInterval.DAILY
        }
        binding.chipRepeatWeekly.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) selectedRepeat = RepeatInterval.WEEKLY
        }
    }

    private fun setupClickListeners() {
        binding.buttonSelectDate.setOnClickListener { showDatePickerDialog() }
        binding.buttonSelectTime.setOnClickListener { showTimePickerDialog() }
        binding.buttonSaveTask.setOnClickListener { saveTask() }
    }

    private fun checkForEditIntent() {
        taskId = intent.getIntExtra(EXTRA_TASK_ID, -1)
        isEditing = taskId != -1

        if (isEditing) {
            binding.textViewAddTitle.text = getString(R.string.edit_task_title)
            binding.buttonSaveTask.text = getString(R.string.update_task)
            taskViewModel.allTasks.observe(this) { tasks ->
                val task = tasks.firstOrNull { it.id == taskId } ?: return@observe
                binding.editTextTitle.setText(task.title)
                binding.editTextDescription.setText(task.description)
                binding.editTextTags.setText(task.tags)
                selectedCategory = task.categoryId
                selectedPriority = task.priority
                selectedRepeat = task.repeatInterval
                selectedDateCalendar.timeInMillis = task.dueDate
                binding.switchReminder.isChecked = task.reminderEnabled
                binding.switchReminder.isEnabled = task.reminderEnabled
                selectCategoryChip(task.categoryId)
                selectChip(binding.chipGroupPriority, task.priority.toString())
                selectChip(binding.chipGroupRepeat, task.repeatInterval.toString())
                updateDateAndTimeViews()
            }
        } else {
            binding.textViewAddTitle.text = getString(R.string.add_task_title)
            binding.buttonSaveTask.text = getString(R.string.save_task)
            binding.chipCategoryGeneral.isChecked = true
            binding.chipPriorityMedium.isChecked = true
            binding.chipRepeatNone.isChecked = true
        }
    }

    private fun selectChip(chipGroup: com.google.android.material.chip.ChipGroup, value: String) {
        for (i in 0 until chipGroup.childCount) {
            val chip = chipGroup.getChildAt(i) as? Chip ?: continue
            if (chip.text.toString() == value) {
                chip.isChecked = true
            }
        }
    }

    private fun showDatePickerDialog() {
        val datePicker = DatePickerDialog(
            this,
            { _, year, month, dayOfMonth ->
                selectedDateCalendar.set(Calendar.YEAR, year)
                selectedDateCalendar.set(Calendar.MONTH, month)
                selectedDateCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                updateDateAndTimeViews()
            },
            selectedDateCalendar.get(Calendar.YEAR),
            selectedDateCalendar.get(Calendar.MONTH),
            selectedDateCalendar.get(Calendar.DAY_OF_MONTH)
        )
        datePicker.show()
    }

    private fun showTimePickerDialog() {
        val timePicker = TimePickerDialog(
            this,
            { _, hourOfDay, minute ->
                selectedDateCalendar.set(Calendar.HOUR_OF_DAY, hourOfDay)
                selectedDateCalendar.set(Calendar.MINUTE, minute)
                updateDateAndTimeViews()
            },
            selectedDateCalendar.get(Calendar.HOUR_OF_DAY),
            selectedDateCalendar.get(Calendar.MINUTE),
            true
        )
        timePicker.show()
    }

    private fun updateDateAndTimeViews() {
        binding.textViewSelectedDate.text = DateUtils.formatFull(selectedDateCalendar.timeInMillis)
        binding.textViewSelectedTime.text = DateUtils.formatTime(selectedDateCalendar.timeInMillis)
    }

    private fun saveTask() {
        val title = binding.editTextTitle.text.toString().trim()
        val description = binding.editTextDescription.text.toString().trim()
        val tags = binding.editTextTags.text.toString().trim()

        if (title.isEmpty()) {
            binding.editTextTitle.error = getString(R.string.task_title_required)
            binding.editTextTitle.requestFocus()
            return
        }

        val reminderEnabled = binding.switchReminder.isChecked
        val dueDate = selectedDateCalendar.timeInMillis

        if (reminderEnabled && dueDate < System.currentTimeMillis()) {
            MaterialAlertDialogBuilder(this)
                .setTitle(R.string.past_date_warning)
                .setPositiveButton(R.string.yes) { _, _ -> persistTask(title, description, tags, reminderEnabled, dueDate) }
                .setNegativeButton(R.string.cancel, null)
                .show()
        } else {
            persistTask(title, description, tags, reminderEnabled, dueDate)
        }
    }

    private fun persistTask(
        title: String,
        description: String,
        tags: String,
        reminderEnabled: Boolean,
        dueDate: Long
    ) {
        if (isEditing) {
            taskViewModel.allTasks.value?.firstOrNull { it.id == taskId }?.let { existing ->
                val updated = existing.copy(
                    title = title,
                    description = description,
                    dueDate = dueDate,
                    categoryId = selectedCategory,
                    priority = selectedPriority,
                    reminderEnabled = reminderEnabled,
                    repeatInterval = selectedRepeat,
                    tags = tags
                )
                taskViewModel.update(updated)
                if (reminderEnabled) {
                    alarmHelper.setAlarm(
                        taskId, title, description, dueDate, selectedRepeat
                    )
                } else {
                    alarmHelper.cancelAlarm(taskId)
                }
                Toast.makeText(this, R.string.task_updated, Toast.LENGTH_SHORT).show()
            }
        } else {
            taskViewModel.insert(
                Task(
                    title = title,
                    description = description,
                    dueDate = dueDate,
                    categoryId = selectedCategory,
                    priority = selectedPriority,
                    reminderEnabled = reminderEnabled,
                    repeatInterval = selectedRepeat,
                    tags = tags
                )
            ) { id ->
                if (reminderEnabled) {
                    alarmHelper.setAlarm(
                        id.toInt(), title, description, dueDate, selectedRepeat
                    )
                }
            }
            Toast.makeText(this, R.string.task_saved, Toast.LENGTH_SHORT).show()
        }

        finish()
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }
}